### PR TITLE
[WIP] Render system, materials and model loading changes

### PR DESCRIPTION
**Pull request by keptsecret**
Implementing materials and re-organizing loading models  
Will depend on #9 for ECS implementation

### CODE REVIEW

File: src/engine/include/engine/RenderSystem.hpp [hunk 1]
```diff
@@ -0,0 +1,56 @@
+#ifndef LEI3D_RENDERSYSTEM_HPP
```
Comment (kevinsadi) [inline src/engine/include/engine/RenderSystem.hpp:1]: When we merge this with Logan's stuff we'd probably want to use `#pragma once`

File: src/engine/Mesh.cpp [hunk 2]
```diff
@@ -31,21 +31,24 @@ namespace lei3d
     
         glBindVertexArray(VAO);
         glBindBuffer(GL_ARRAY_BUFFER, VBO);
-        glBufferData(GL_ARRAY_BUFFER, vertices.size() * sizeof(Vertex), &vertices[0], GL_STATIC_DRAW);  
+        glBufferData(GL_ARRAY_BUFFER, vertices.size() * sizeof(Vertex), vertices.data(), GL_STATIC_DRAW);
```
Comment (kevinsadi) [inline src/engine/Mesh.cpp:34]: nice, good change

File: data/shaders/forward.frag [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+//
+//        vec3 bump = normalize(vec3(dxy * material.bump_scale / texelSize, 1.0));
+//        bump = normalize(bump * 0.5 + 0.5);
+//        N = normalize(TBN * bump);
+//    }
+
+    vec3 lightColor = dirLight.color * dirLight.intensity;
+    vec3 outColor = lightColor * albedo.rgb;
+    float NdotL = dot(dirLight.direction, N);
+    if (NdotL > 0.95) {
```
Comment (kevinsadi) [inline data/shaders/forward.frag]: This toon shader looks really nice, good job.

File: data/shaders/forward.frag [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    } else if (NdotL > 0.5) {
+        outColor = outColor * 0.75;
+    } else if (NdotL > 0.25) {
+        outColor = outColor * 0.4;
+    } else {
+        outColor = outColor * 0.1;
+    }
+    FragOut = outColor;
+
+    // TODO: temporary, mock position and radius of color source
```
Comment (kevinsadi) [inline data/shaders/forward.frag:86]: I assume we would pass this to the shader once we add a color source. I know we are currently planning on just having a single directional light from the sun, but how would we support more color sources?

File: data/shaders/forward.frag [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+struct DirLight {
+    vec3 direction;
+    vec3 color;
+    float intensity;
+};
+
+in vec3 FragPos;
+in vec3 Normal;
+in vec2 TexCoords;
+in mat3 TBN;
```
Comment (kevinsadi) [inline data/shaders/forward.frag:40]: Where is TBN being used? What does it do?

File: data/shaders/screenspace_quad.vert [hunk 6]
```diff
@@ -0,0 +1,11 @@
+#version 330 core
+
+out vec2 TexCoords;
+
+void main() {
+    vec2 position = vec2(gl_VertexID % 2, gl_VertexID / 2) * 4.0 - 1;
```
Comment (kevinsadi) [inline data/shaders/screenspace_quad.vert:6]: What is this vaguely cryptic equation doing?

File: src/engine/RenderSystem.cpp [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+    scwidth = width;
+    scheight = height;
+
+    forwardShader = Shader("./data/shaders/forward.vert", "./data/shaders/forward.frag");
+    postprocessShader = Shader("./data/shaders/screenspace_quad.vert", "./data/shaders/postprocess.frag");
+
+    glGenVertexArrays(1, &dummyVAO);
+
+    glGenFramebuffers(1, &FBO);
+    glBindFramebuffer(GL_DRAW_FRAMEBUFFER, FBO);
```
Comment (kevinsadi) [inline src/engine/RenderSystem.cpp:18]: Never saw the use of a frame buffer object, what a great example for it though. Learned a lot from this PR

File: src/engine/RenderSystem.cpp [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+    glGenTextures(1, &saturationMask);
+    glGenTextures(1, &depthStencilTexture);
+    glGenTextures(1, &finalTexture);
+
+    // lighting pass
+    glBindTexture(GL_TEXTURE_2D, rawTexture);
+    glTexImage2D(GL_TEXTURE_2D, 0, GL_RGB32F, width, height, 0, GL_RGB, GL_FLOAT, nullptr);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
+    glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D, rawTexture, 0);
```
Comment (kevinsadi) [inline src/engine/RenderSystem.cpp:30]: Do these load the output of the color attachments to the texture that you minded in glBindTexture?

Comment (kevinsadi) [reply src/engine/RenderSystem.cpp:30]: Or does it also attach it the texture to the Frame buffer object?

File: src/engine/RenderSystem.cpp [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+    glBindTexture(GL_TEXTURE_2D, depthStencilTexture);
+    glTexImage2D(GL_TEXTURE_2D, 0, GL_DEPTH32F_STENCIL8, width, height, 0, GL_DEPTH_STENCIL, GL_FLOAT_32_UNSIGNED_INT_24_8_REV, nullptr);   // probably won't need stencil
+    glFramebufferTexture2D(GL_FRAMEBUFFER, GL_DEPTH_STENCIL_ATTACHMENT, GL_TEXTURE_2D, depthStencilTexture, 0);
+
+    // post process pass target
+    glBindTexture(GL_TEXTURE_2D, finalTexture);
+    glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, width, height, 0, GL_RGB, GL_FLOAT, nullptr);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
+    glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT2, GL_TEXTURE_2D, finalTexture, 0);
```
Comment (kevinsadi) [inline src/engine/RenderSystem.cpp:49]: Err, I understand that these steps are certainly procedural in nature, but I don't understand how these gl calls relate to each other procedurally. Very very cool and clever though.

**Review by kevinsadi — COMMENTED** [review]
So far looks great! Excited to get it merged with Logan's stuff

**Comment by keptsecret** [comment]
You can test the renderer by checking out this branch https://github.com/keptsecret/lei3d/tree/test-renderer-legacy
This is before #9 is merged with main, and will cause plenty of conflicts if merged

**Comment by keptsecret** [comment]
Since the project restructuring, moved all progress into #19. This PR will be closed

**Pull request closed without merging**